extensions/spark/kyuubi-extension-spark-common/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxScanStrategy.scala

```diff
@@ -0,0 +1,293 @@
+    if (maxScanPartitions > 0 || maxFileSize > 0) {
+      checkScan(plan, maxScanPartitions, maxFileSize)
+    }
+    Nil
+  }
+
+  private def checkScan(plan: LogicalPlan, maxScanPartitions: Int, maxFileSize: Long): Unit = {
+    plan match {
+      case ScanOperation(_, _, relation: HiveTableRelation) =>
+        if (relation.isPartitioned) {
+          relation.prunedPartitions match {
+            case Some(prunedPartitions) =>
+              if (maxScanPartitions > 0 && prunedPartitions.size > maxScanPartitions) {
+                throw new MaxPartitionExceedException(
+                  s"""
+                     |SQL job scan hive partition: ${prunedPartitions.size}
+                     |exceed restrict of hive scan maxPartition $maxScanPartitions
+                     |You should optimize your SQL logical according partition structure
+                     |or shorten query scope such as p_date, detail as below:
+                     |Table: ${relation.tableMeta.qualifiedName}
+                     |Owner: ${relation.tableMeta.owner}
+                     |Partition Structure: ${relation.partitionCols.map(_.name).mkString(", ")}
+                     |""".stripMargin)
+              }
+              lazy val scanFileSize = prunedPartitions.flatMap(_.stats).map(_.sizeInBytes).sum
+              if (maxFileSize > 0 && scanFileSize > maxFileSize) {
+                throw partTableMaxFileExceedError(
+                  scanFileSize,
+                  maxFileSize,
+                  Some(relation.tableMeta),
+                  prunedPartitions.flatMap(_.storage.locationUri).map(_.toString),
+                  relation.partitionCols.map(_.name))
+              }
+            case _ =>
+              val totalPartitions = session
+                .sessionState.catalog.externalCatalog.listPartitions(
+                  relation.tableMeta.database,
+                  relation.tableMeta.identifier.table)
+              if (maxScanPartitions > 0 && totalPartitions.size > maxScanPartitions) {
+                throw new MaxPartitionExceedException(
+                  s"""
+                     |Your SQL job scan a whole huge table without any partition filter,
+                     |You should optimize your SQL logical according partition structure
+                     |or shorten query scope such as p_date, detail as below:
+                     |Table: ${relation.tableMeta.qualifiedName}
+                     |Owner: ${relation.tableMeta.owner}
+                     |Partition Structure: ${relation.partitionCols.map(_.name).mkString(", ")}
+                     |""".stripMargin)
+              }
+              lazy val scanFileSize = totalPartitions.flatMap(_.stats).map(_.sizeInBytes).sum
+              if (maxFileSize > 0 && scanFileSize > maxFileSize) {
+                throw new MaxFileSizeExceedException(
+                  s"""
+                     |Your SQL job scan a whole huge table without any partition filter,
```

Comment:
shall we change the hint of exception ? user may have a valid partition filter, but the file size is still large enough

Reply:
> user may have a valid partition filter, but the file size is still large enough

Here matches the table without partition filtering.

```
relation.prunedPartitions match {
  ...
  case _ >  // This means the prunedPartitions is None
```